File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index_map = {}  # last index of each character
        max_length = 0
        start = 0  # Start of the sliding window

        for end, char in enumerate(s):
            if char in char_index_map and char_index_map[char] >= start:
                # If the character is repeated and it's within the current window
                start = char_index_map[char] + 1  # Move the start of the window

            # Update the last index of the character
            char_index_map[char] = end

            current_length = end - start + 1
            max_length = max(max_length, current_length)

        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        Map<Character, Integer> map = new HashMap<>();

        int maxLen = 0;
        int startIdx = 0;
        for (int i = 0; i < n; i++){
            if (map.containsKey(s.charAt(i))){
                int nextIdx = map.get(s.charAt(i)) + 1;
                startIdx = Math.max(startIdx, nextIdx);
            } 
            map.put(s.charAt(i), i);
            maxLen = Math.max(maxLen, i - startIdx + 1);
            // System.out.println(map);
            // System.out.println(s.charAt(i));
            // System.out.println(maxLen);
            // System.out.println();
        }    


        return maxLen;
    }
}